### PR TITLE
HTTP/2: yet another PR

### PR DESCRIPTION
**Pull request by Kriechi**
more things required for Pathod

* simplified `human_readable` for smaller output
* add elliptic curve support
     * might needs a more sophisticated approach - but works for now
* improve Header/Continuation-Frame handling
* return the `stream_id` from a request, to be used for creating a response
* properly set the `:authority` header

### CODE REVIEW

File: netlib/tcp.py [hunk 1]
```diff
@@ -529,7 +527,7 @@ def get_alpn_proto_negotiated(self):
         if OpenSSL._util.lib.Cryptography_HAS_ALPN and self.ssl_established:
             return self.connection.get_alpn_proto_negotiated()
         else:
-            return None
+            return ""
```
Comment (mhils) [inline netlib/tcp.py]: Why this change?

File: netlib/http2/protocol.py [hunk 2]
```diff
@@ -48,16 +49,19 @@ def check_alpn(self):
...
 
-    def _receive_settings(self):
-        frm = frame.Frame.from_file(self.tcp_handler.rfile, self)
-        assert isinstance(frm, frame.SettingsFrame)
-        self._apply_settings(frm.settings)
+    def _receive_settings(self, hide=False):
+        while True:
+            frm = self.read_frame(hide)
+            if isinstance(frm, frame.SettingsFrame):
+                break
```
Comment (mhils) [inline netlib/http2/protocol.py]: This function looks weird. We just read frames and discard them until we stumble upon a settings frame? Don't we swallow possibly relevant packets by this?

File: netlib/http2/protocol.py [hunk 3]
```diff
@@ -93,30 +95,37 @@ def next_stream_id(self):
...
-        raw_bytes = frame.to_bytes()
+    def send_frame(self, frm, hide=False):
+        raw_bytes = frm.to_bytes()
         self.tcp_handler.wfile.write(raw_bytes)
         self.tcp_handler.wfile.flush()
+        if not hide and self.dump_frames:  # pragma no cover
+            print(frm.human_readable(">>"))
 
-    def read_frame(self):
+    def read_frame(self, hide=False):
```
Comment (mhils) [inline netlib/http2/protocol.py]: What's the point of the `hide` param?

File: netlib/http2/protocol.py [hunk 4]
```diff
@@ -127,12 +136,16 @@ def _create_headers(self, headers, stream_id, end_stream=True):
...
             flags=flags,
             stream_id=stream_id,
-            header_block_fragment=header_block_fragment).to_bytes()
-        return [bytes]
+            header_block_fragment=header_block_fragment)
+
+        if self.dump_frames:  # pragma no cover
+            print(frm.human_readable(">>"))
+
+        return [frm.to_bytes()]
```
Comment (mhils) [inline netlib/http2/protocol.py]: It seems weird that we return an array here. Maybe just wrap the call in `itertools.chain`?

File: netlib/http2/protocol.py (hunk 2 above)
Comment (Kriechi) [inline netlib/http2/protocol.py]: The RFC states that Settings must be processed immediately and the ACK frame must be sent ASAP.
Yes - it's kinda strange - but it's the easiest solution without introducing a global queue and request/response pipelines... - a full-blown HTTP/2 implementation should probably use a more sophisticated approach.

File: netlib/http2/protocol.py (hunk 3 above)
Comment (Kriechi) [inline netlib/http2/protocol.py]: Pathod/Pathoc has a new CLI argument `--http2-framedump` which will output all frames (incoming & outgoing).
But usually I'm not interested in the connection preface frames - as they are always the same.
Therefore I want to hide only those frame with this argument set to `True`.

File: netlib/http2/protocol.py (hunk 4 above)
Comment (Kriechi) [inline netlib/http2/protocol.py]: should yield the same result?

File: netlib/tcp.py (hunk 1 above)
Comment (Kriechi) [inline netlib/tcp.py]: `self.connection.get_alpn_proto_negotiated()` will return `""` if NO application layer protocol was negotiated.
This change will make this wrapper function to behave in the same way. Otherwise we would have two different ways to represent the state: "<no alpn performed>"

File: netlib/http2/protocol.py (hunk 4 above)
Comment (mhils) [inline netlib/http2/protocol.py]: Why not return the bytes here and then use a plain simple array in place of itertools.chain?

Comment (Kriechi) [inline netlib/http2/protocol.py]: It might not only be a single frame we want to return here.
Once we support ContinuationFrames we want to return a list of frames. 
Therefore using an array here is future-proof.

File: netlib/http2/protocol.py (hunk 2 above)
Comment (cortesi) [inline netlib/http2/protocol.py]: I'm merging, but I'd like the API to give us some way to access discarded frames. In the mitmproxy UIs, we'll probably have something to toggle between a frame view and an HTTP view, and we'll need the frame view to be complete. If this means keeping state on something stored on the connection, we should bake that into this low-level API.
